Create the category service

BrainTireFire opened the pull request:
Summary:

- Created the CategoryService to manage category-related operations such as creating, updating, finding, and deleting categories.
- Implemented methods to handle pagination with proper DTO mapping to avoid the N+1 problem.
- Introduced short and full DTOs for efficient category data transfer, providing flexibility based on the required response.
- Handled cascading deletions: When deleting a category, the associated games are also removed to maintain data integrity.
- Added PaginationUtils to centralize and simplify the logic for paginated responses.

BrainTireFire committed:
Create the category service

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/category/Category.java:
```diff
@@ -23,6 +24,6 @@ public class Category extends BaseEntity {
     private String description;
 
-    @OneToMany(mappedBy = "category")
+    @OneToMany(mappedBy = "category", cascade = CascadeType.ALL, orphanRemoval = true)
```
Is CascadeType.ALL really needed in here?
Shouldn't we use precise cascadings?

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/category/CategoryMapper.java:
```diff
@@ -0,0 +1,59 @@
+                .build();
+    }
+
+    public CategoryResponseShort toCategoryResponseShort(Category category) {
```
rename the DTO to ShortCategoryResponse

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/category/CategoryMapper.java:
```diff
@@ -0,0 +1,59 @@
+                .build();
+    }
+
+    public CategoryResponseFull toCategoryResponseFull(Category category) {
```
I don't see the need for this full category response for the following reasons:

1. Performance issues: the games are not paginated, image having thousands of games in one category
2. If we want the games that belong to that category, we need to fetch using the appropriate API

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/category/CategoryMapper.java:
```diff
@@ -0,0 +1,59 @@
+                .collect(Collectors.toList());
+    }
+
+    private GameResponse toGameResponse(Game game) {
```
Single Responsibility?

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/category/CategoryRepository.java:
```diff
@@ -1,7 +1,31 @@
 public interface CategoryRepository extends JpaRepository<Category, String> {
 
+    // To avoid N+1 queries problem, we fetch the games and platforms of the category
+    @Query("""
```
As mentioned above, we should not return categories with games.

Also this query is not really needed (even in other cases) because when you use `categoryRepository.findAll(pageable)` and make the method `@Transactional` you should have access to the Games that are in that category

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/utils/PaginationUtils.java:
```diff
@@ -0,0 +1,30 @@
+
+    private PaginationUtils() { }
+
+    public static <T, R> PageResponse<R> buildPageResponse(Page<T> page, Function<T, R> mapper) {
```
I like this part in here.
This saves us from writing the same code many times.

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/category/CategoryService.java:
```diff
@@ -0,0 +1,89 @@
+        final Category category = categoryRepository.findById(categoryId)
+                .orElseThrow(() -> {
+                    log.warn("Category with id {} does not exist", categoryId);
+                    return new RuntimeException("Category with id " + categoryId + " does not exist");
```
Better throw a dedicated exception

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/category/CategoryService.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    public CategoryResponseShort findCategoryById(String categoryId) {
+        final Category category = categoryRepository.findById(categoryId)
```
The code is fine for me.
As explained in the session, I would `findById()
.map(mapperMethod)
.orElseThrow(()->new EntityNotFoundException("the proper message"));`

Then in the GlobalExceptionHandler (we will implement it later) I can catch the exception, log it and the return a BAD_REQUEST

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/category/CategoryService.java:
```diff
@@ -0,0 +1,89 @@
+        return categoryMapper.toCategoryResponseShort(category);
+    }
+
+    public PageResponse<CategoryResponseFull> findAllCategoriesWithGamesAndPlatforms(int page, int size) {
```
to be removed

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/category/CategoryService.java:
```diff
@@ -0,0 +1,89 @@
+
+    public String saveCategory(final CategoryRequest categoryRequest) {
+
+        if (categoryRepository.existsByName(categoryRequest.name())) {
```
We should check the lower case (it should not be case sensitive)
EG: The Witcher iii === the Witcher III

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/category/CategoryService.java:
```diff
@@ -0,0 +1,89 @@
+                    return new RuntimeException("Category with id " + categoryId + " does not exist");
+                });
+
+        final Category updatedCategory = categoryMapper.toCategory(categoryRequest);
```
We need to check of we have the new category name already in the DB

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/category/CategoryService.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    public void deleteCategory(String categoryId) {
+        final Category category = categoryRepository.findById(categoryId)
```
I would recommend that you check the latest video where we implemented the delete method

ali-bouali commented on game-store-api/src/main/java/com/alibou/store/game/GameResponse.java:
```diff
@@ -17,6 +17,7 @@ public class GameResponse {
 
     private String id;
     private String name;
```
I would rename the name attr to title (this is what we did the last time)

ali-bouali reviewed (commented):
First of all, thank you for the contribution and the effort.
I really like the and what you suggested.
Please review my comments, make adjustments and I will re-review it again and if everything is fine, I will merge it

PS: Make sure that all the conflicts are resolved

BrainTireFire committed:
Correct the comment from Ali Bouali

BrainTireFire committed:
Merge branch 'main' into main

BrainTireFire commented:
Hi @ali-bouali 

Sorry I couldn't make it to the last session to discuss this code. I've made the changes you mentioned in the comments.

Recap:

1. Removed the full category response.
2. Added existsByNameIgnoreCase.
3. Corrected the exception errors.
4. Added checks to the update method.
5. Rewrote the remove method:
                  - Added a check: if games exist, it returns an error.
6. Make whole service @Transactional

BrainTireFire committed:
Make the whole service @Transactional
